### PR TITLE
[Zen2] Integrate FollowerChecker with Coordinator

### PR DESCRIPTION
**Pull request by DaveCTurner**
This change ensures that the leader node periodically checks that its followers
are healthy, and that they are removed from the cluster if not.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java [hunk 1]
```diff
@@ -203,15 +232,27 @@ void stabilise(long stabilisationTime) {
                 }
 
                 if (deterministicTaskQueue.hasDeferredTasks() == false) {
-                    break; // TODO when fault detection is enabled this should be removed, as there should _always_ be deferred tasks
+                    // A 1-node cluster has no need for fault detection etc so will eventually run out of things to do.
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java]: 1 node cluster = best cluster 💃

**Review by ywelsch — COMMENTED** [review]
I've left some smaller comments. Looks very good already.

File: server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java [hunk 2]
```diff
@@ -378,6 +452,10 @@ assert getStateForMasterService().nodes().getMasterNodeId() != null
                     || getStateForMasterService().term() != getCurrentTerm() :
                     getStateForMasterService();
                 assert leaderCheckScheduler == null : leaderCheckScheduler;
+
+                if (publicationInProgress() || getLastCommittedState().map(s -> s.term() == getCurrentTerm()).orElse(false)) {
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java]: Ok, assertions are nicer in 916f1f56531bf2e9d5b50f20ce7a696610dfb158.

**Review by ywelsch — APPROVED** [review]
LGTM (one nit)

**Comment by DaveCTurner** [comment]
@elasticmachine test this please

**Pull request merged**